ykrt/src/mt.rs

```diff
@@ -41,14 +45,62 @@ const PHASE_COUNTING: usize = 0b10; // The value specifies how many times we've
 /// mechanism is for every bytecode or AST node to have its own `Location` (even for bytecodes or
 /// nodes that can't be control points).
 #[derive(Debug)]
-pub struct Location {
-    pack: AtomicUsize,
+pub struct Location<I> {
+    // A Location is a state machine which operates as follows (where PHASE_COUNTING is the start
+    // state):
+    //
+    //               ┌────────────────────────────────────────┐
+    //               │                                        │ ─────────────┐
+    //   reprofile   │             PHASE_COUNTING             │              │
+    //  ┌──────────▶ │                                        │ ◀────────────┘
+    //  │            └────────────────────────────────────────┘    increment
+    //  │              │                     ▲         ▲           count
+    //  │              │ start tracing       │ abort   │ abort
+    //  │              ▼                     │         │
+    //  │            ┌────────────────────┐  │         │
+    //  │            │   PHASE_TRACING    │ ─┘         │
+    //  │            └────────────────────┘            │
+    //  │              │                               │
+    //  │              │ start compiling trace         │
+    //  │              │ in thread                     │
+    //  │              ▼                               │
+    //  │            ┌────────────────────┐            │
+    //  |            |                    | ───────────┘
+    //  │            │                    │
+    //  │            │  PHASE_COMPILING   │ ──────┐
+    //  │            │                    │       │ still compiling in thread
+    //  │            │                    │ ◀─────┘
+    //  │            └────────────────────┘
+    //  │              │
+    //  │              │ trace compiled
+    //  │              ▼
+    //  │            ┌────────────────────┐
+    //  │            │                    │ ──────┐
+    //  │            │   PHASE_COMPILED   │       │
+    //  └─────────── │                    │ ◀─────┘
+    //               └────────────────────┘
+    //
+    // We hope that a Location soon reaches PHASE_COMPILED (aka "the happy state") and stays there.
+    state: AtomicUsize,
+    phantom: PhantomData<I>,
 }
 
-impl Location {
+impl<I> Location<I> {
     pub fn new() -> Self {
         Self {
-            pack: AtomicUsize::new(PHASE_COUNTING),
+            state: AtomicUsize::new(PHASE_COUNTING),
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<I> Drop for Location<I> {
+    fn drop(&mut self) {
+        let lp = self.state.load(Ordering::Acquire);
```

Comment:
Can this be relaxed ordering?

Reply:
Good point! Fixed in https://github.com/softdevteam/yk/pull/171/commits/d9bfa2e563763dfcca8dc374af2e52adfbdb9fbe.